### PR TITLE
ambient: fix stale endpoint info

### PR DESCRIPTION
**Pull request by howardjohn**
**Please provide a description of this PR:**

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -2262,6 +2263,16 @@ func (ps *PushContext) ServiceAccounts(hostname host.Name, namespace string, por
 	}]
 }
 
+func (ps *PushContext) SupportsTunnel(ip string) bool {
+	infos, _ := ps.ambientIndex.PodInformation(sets.New(types.NamespacedName{Name: ip}))
+	for _, p := range infos {
+		if p.Protocol == workloadapi.Protocol_HTTP {
```
Comment (linsun) [inline pilot/pkg/model/push_context.go:2269]: Q - I saw we no longer using the redirection annotation to detect if it is in ambient.  How would protocol http here help us detect if supports tunnel or not?

Comment (howardjohn) [reply pilot/pkg/model/push_context.go:2269]: I updated the PR description with a bunch of info.

tl;dr Protocol_HTTP is derived from annotation, we are just using the derived info instead of the source (which was a leaky abstraction)

Comment (linsun) [reply pilot/pkg/model/push_context.go:2269]: Thanks, very helpful!

**Comment by linsun** [comment]
I tried the change and verified the L4 authz policy is fixed from ingress gateway to productpage.

I also dumped out the EDS config for istio ingress gateway and observed the change:

```
    {
     "endpoint_config": {
      "@type": "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment",
      "cluster_name": "outbound-tunnel",
      "endpoints": [
       {
        "locality": {},
        "lb_endpoints": [
         {
          "endpoint": {
           "address": {
            "socket_address": {
             "address": "10.244.1.9",
             "port_value": 15008
            }
           },
           "health_check_config": {},
           "hostname": "outbound-tunnel10.244.1.9:15008"
          },
          "health_status": "HEALTHY",
          "load_balancing_weight": 1
         }
        ]
       }
      ],
      "policy": {
       "overprovisioning_factor": 140
      }
     }
    },
```

**Review by linsun — APPROVED** [review]
Reviewed the endpoint builder change, LGTM

**Review by GregHanson — APPROVED** [review]
added an integration test for the original `updateEndpointsOnWaypointChange` issue [here](https://github.com/istio/istio/pull/43699), verified John's changes work for this one as well

**Comment by linsun** [comment]
nice @GregHanson !!

**Review by hzxuzhonghu — APPROVED** [review]
